src/input.rs

```diff
@@ -266,14 +300,51 @@ impl<'a> Input<'a> {
         Ok(std::str::from_utf8(out.as_slice()).unwrap().to_string())
     }
 
+    fn validate(&mut self) -> io::Result<()> {
+        let validation = (self.validator)(&self.input);
+        if validation.is_err() {
+            self.err = Some(validation.unwrap_err().to_string());
+        } else {
+            self.err = None;
+        }
+        Ok(())
+    }
+
     fn set_cursor(&mut self) -> io::Result<()> {
         if !self.placeholder.is_empty() && self.input.is_empty() {
             self.term
                 .move_cursor_left(self.placeholder.chars().count())?;
         } else {
             self.term.move_cursor_left(self.input.chars().count())?;
         }
-        self.term.move_cursor_right(self.cursor)?;
+
+        // if there is an error, move the cursor up from error message and right to the input
+        match self.err {
```

Comment:
cursor handling is a nightmare. i wonder if i'm doing something wrong here but without being able to get the initial cursor position i don't see another way than working backwards from end of the output.

Reply:
yikes.

Reply:
that's all I have to say lol